### PR TITLE
Add protocol-related metrics

### PR DESCRIPTION
**Pull request by LukasGasior1**

### CODE REVIEW

File: core-rust/state-manager/src/metrics.rs [hunk 1]
```diff
@@ -330,6 +344,202 @@ impl CommittedTransactionsMetrics {
...
+                ),
+                &["readiness_signal_name"],
+            )
+            .registered_at(registry),
+            pending_update_threshold_required_ratio: GaugeVec::new(
+                opts(
+                    "protocol_update_pending_threshold_required_ratio",
+                    "Required readiness ratio for the given protocol update threshold.",
+                ),
+                &["protocol_version_name", "readiness_signal_name", "threshold_index"],
```
Comment (dhedey) [inline core-rust/state-manager/src/metrics.rs:363]: Maybe we should include a `threshold_epochs` label too in these :)

File: core-rust/state-manager/src/metrics.rs [hunk 2]
```diff
@@ -330,6 +344,202 @@ impl CommittedTransactionsMetrics {
...
+                .registered_at(registry),
+            pending_update_threshold_required_consecutive_epochs: IntGaugeVec::new(
+                opts(
+                    "protocol_update_pending_threshold_required_consecutive_epochs",
+                    "Required number of consecutive epochs of support for the given protocol update threshold.",
+                ),
+                &["protocol_version_name", "readiness_signal_name", "threshold_index"],
+            )
+                .registered_at(registry),
+            pending_update_threshold_matching_consecutive_epochs: IntGaugeVec::new(
```
Comment (dhedey) [inline core-rust/state-manager/src/metrics.rs:374]: Ohh nice idea! :)

We should have a `pending_update_threshold_required_consecutive_epochs` as well as a `pending_update_threshold_current_consecutive_epochs` so we can draw a graph and check the current vs the required.

File: core-rust/state-manager/src/metrics.rs [hunk 3]
```diff
@@ -330,6 +344,202 @@ impl CommittedTransactionsMetrics {
...
+                    "protocol_update_pending_upper_bound_epoch",
+                    "Upper bound epoch for the given protocol update (exclusive)",
+                ),
+                &["protocol_version_name", "readiness_signal_name"],
+            )
+                .registered_at(registry),
+            enacted_protocol_update_state_version: IntGaugeVec::new(
+                opts(
+                    "protocol_update_enacted_state_version",
+                    "State version of enacted protocol update",
```
Comment (dhedey) [inline core-rust/state-manager/src/metrics.rs:401]: Maybe worth clarifying it's the state version of the init of the protocol update?

**Review by dhedey — APPROVED** [review]
Have just reviewed the definitions, not the implementation, but it looks good :). Have a few suggestions!

File: core-rust/state-manager/src/metrics.rs (hunk 2 above)
Comment (LukasGasior1) [reply core-rust/state-manager/src/metrics.rs:374]: yup, that's the "current" (I'll rename it)

File: core-rust/state-manager/src/metrics.rs (hunk 1 above)
Comment (LukasGasior1) [reply core-rust/state-manager/src/metrics.rs:363]: hmh it's already there as a separate gauge, not sure why we would need it as a label?

Comment (dhedey) [reply core-rust/state-manager/src/metrics.rs:363]: Fair - I was just thinking if you want a single graph of these lines, along with the current ratios, being able to label with the threshold epochs is probably more informative than the threshold index

Comment (LukasGasior1) [reply core-rust/state-manager/src/metrics.rs:363]: ah, gotcha :+1: okay, I'll add it

**Pull request merged**